util.go

```diff
@@ -0,0 +1,128 @@
+package jsonserialization
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+)
+
+type numericRange struct {
+	min          float64
+	max          float64
+	allowDecimal bool
+}
+
+var (
+	numericTypeRanges = map[reflect.Kind]numericRange{
+		reflect.Int8:    {math.MinInt8, math.MaxInt8, false},
+		reflect.Uint8:   {0, math.MaxUint8, false},
+		reflect.Int16:   {math.MinInt16, math.MaxInt16, false},
+		reflect.Uint16:  {0, math.MaxUint16, false},
+		reflect.Int32:   {math.MinInt32, math.MaxInt32, false},
+		reflect.Uint32:  {0, math.MaxUint32, false},
+		reflect.Int64:   {math.MinInt64, math.MaxInt64, false},
+		reflect.Uint64:  {0, math.MaxUint64, false},
+		reflect.Float32: {-math.MaxFloat32, math.MaxFloat32, true},
+		reflect.Float64: {-math.MaxFloat64, math.MaxFloat64, true},
+	}
+)
+
+type number interface {
+	int | int8 | uint8 | int16 | uint16 | int32 | uint32 | int64 | uint64 | float32 | float64
+}
+
+// isCompatible checks if the value is compatible with the type tp.
+// It intentionally excludes checking if types are pointers to allow for possibility.
+func isCompatible(value interface{}, tp reflect.Type) bool {
+	// Can't join with lower, number types are always "convertible" just not losslessly.
+	if isNumericType(value) && isNumericType(tp) {
+		//NOTE: no need to check if number is compatible with another, always yes, just overflows
+		//Check if number value is TRULY compatible
+		return isCompatibleInt(value, tp)
+	}
+
+	return reflect.TypeOf(value).ConvertibleTo(tp)
+}
+
+// as converts the value to the type T.
+func as[T any](in interface{}, out T) error {
+	// No point in trying anything if already nil
+	if in == nil {
+		return nil
+	}
+
+	// Make sure nothing is a pointer
+	valValue := reflect.ValueOf(in)
```

Comment:
while I understand the benefits of the approach for numeric types and make sure a conversion is possible/desirable, I'm a bit worried about the potential performance impact of using reflection for the other types.
Have we considered a hybrid approach here? where string/bool/etc... would rely on type assertions and numeric types would rely on this util method?

Reply:
just to make sure I am understanding, your recommendation would be to have only numeric types implement the as function and the "other" types only use type assertions?
ex:
numeric
```golang
func (n *JsonParseNode) GetInt8Value() (*int8, error) {
	if isNil(n) || isNil(n.value) {
		return nil, nil
	}
	var val int8

	if err := as(n.value, &val); err != nil {
		return nil, err
	}

	return &val, nil
}
```
string
```golang
func (n *JsonParseNode) GetStringValue() (*string, error) {
	if isNil(n) || isNil(n.value) {
		return nil, nil
	}
	
       val, ok := n.value.(string)
       if !ok {
          return nil, fmt.Errorf("value '%v' is not compatible with type string", n.value)
       }

	return &val, nil
}
```

Reply:
The error would need to be tweaked since it's a memory address but not really relevant.

Reply:
correct, that's the question I was asking.

Reply:
I don't see any reason not to except for a weird edge case, which shouldn't happen without user error, where the value is set to a double pointer or a string nested in an interface...?

Reply:
for example this would error:
```golang
func main() {

	str := "str"
	strPtr := &str

	node, _ := jsonserialization.NewJsonParseNode([]byte(`"dfsdfds"`))
	node.SetValue(&strPtr)

	val, err := node.GetStringValue()
	if err != nil {
		log.Fatal(err)
	}
	log.Println(val)
}
```
It's just a matter of how much you want to "protect the user" vs what is the performance impact. It might be worth testing it in a dev build, have an intensive object, and compare the two.

Reply:
Thanks for calling this out, this was an oversight on our end, SetValue should never have been public to begin with. I put together a PR to address this. https://github.com/microsoft/kiota-serialization-json-go/pull/145
Do you have other scenarios where you think the change I'm recommending could be problematic?

Reply:
Since `SetValue` will be effectively removed. I can't see any other concerns since only valid JSON will be imported. I can make the change for non-number types.